pkg/ctxlock/ctxlock.go

```diff
@@ -0,0 +1,354 @@
+// Package ctxlock provides a locking mechanism based on context cancellation.
+//
+// Contexts derived from a Locker are canceled when the underlying connection to
+// the lock provider is gone, or when a parent context is canceled.
+//
+// This package makes use of "unsafe" to avoid some allocations, but the "safe"
+// build tag can be provided to use allocating versions of the functions.
+package ctxlock
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/quay/zlog"
+	"go.opentelemetry.io/otel/baggage"
+	"go.opentelemetry.io/otel/label"
+)
+
+// TODO(hank) Specify this algorithm to check its soundness.
+
+// New creates a Locker that will pull connections from the provided pool.
+//
+// The provided context is only used for logging and initial setup. Close must
+// be called to release held resources.
+func New(ctx context.Context, p *pgxpool.Pool) (*Locker, error) {
```

Comment:
More starting a discussion, does it make more sense for the `LockSource` interface definition to be defined here (and add the interface check) and imported? I suppose that would make the dir structure something like:
```
pkg/locksource/locksource.go // interface def
pkg/locksource/ctxlock.go // would need to update to `NewCtxLock`
````
```

Reply:
I think the structure of packages like this runs into a larger issue of structure with `libvuln` and `libindex` themselves: they don't actually use the store and lock abstractions, and other parts of the code do so inconsistently. I think for separating the interface out, we'd want to rework the main objects to be agnostic and use them, as well. Otherwise, it's all an internal detail and the indirection just gets annoying.